security/pkg/k8s/chiron/controller.go

```diff
@@ -0,0 +1,820 @@
+			wc.upsertSecret(saName, scrt.GetNamespace())
+		}
+	}
+}
+
+// scrtUpdated() is the callback function for update event. It handles
+// the certificate rotations.
+func (wc *WebhookController) scrtUpdated(oldObj, newObj interface{}) {
+	scrt, ok := newObj.(*v1.Secret)
+	if !ok {
+		log.Warnf("Failed to convert to secret object: %v", newObj)
+		return
+	}
+	namespace := scrt.GetNamespace()
+	name := scrt.GetName()
+	// Only handle webhook secret update events
+	if !wc.isWebhookSecret(name, namespace) {
+		return
+	}
+
+	certBytes := scrt.Data[CertChainID]
+	cert, err := util.ParsePemEncodedCertificate(certBytes)
+	if err != nil {
+		log.Warnf("Failed to parse certificates in secret %s/%s (error: %v), refreshing the secret.",
+			namespace, name, err)
+		if err = wc.refreshSecret(scrt); err != nil {
+			log.Errora(err)
+		}
+
+		return
+	}
+
+	certLifeTimeLeft := time.Until(cert.NotAfter)
+	certLifeTime := cert.NotAfter.Sub(cert.NotBefore)
+	// TODO(myidpt): we may introduce a minimum gracePeriod, without making the config too complex.
+	// Because time.Duration only takes int type, multiply gracePeriodRatio by 1000 and then divide it.
+	gracePeriod := time.Duration(wc.gracePeriodRatio*1000) * certLifeTime / 1000
+	if gracePeriod < wc.minGracePeriod {
+		log.Warnf("gracePeriod (%v * %f) = %v is less than minGracePeriod %v. Apply minGracePeriod.",
+			certLifeTime, wc.gracePeriodRatio, gracePeriod, wc.minGracePeriod)
+		gracePeriod = wc.minGracePeriod
+	}
+
+	// Refresh the secret if 1) the certificate contained in the secret is about
+	// to expire, or 2) the root certificate in the secret is different than the
+	// one held by the ca (this may happen when the CA is restarted and
+	// a new self-signed CA cert is generated).
+	if certLifeTimeLeft < gracePeriod || !bytes.Equal(wc.getCurCACert(), scrt.Data[RootCertID]) {
+		log.Infof("Refreshing secret %s/%s, either the leaf certificate is about to expire "+
+			"or the root certificate is outdated", namespace, name)
+
+		if err = wc.refreshSecret(scrt); err != nil {
+			log.Errorf("Failed to update secret %s/%s (error: %s)", namespace, name, err)
+		}
+	}
+}
+
+// refreshSecret is an inner func to refresh cert secrets when necessary
+func (wc *WebhookController) refreshSecret(scrt *v1.Secret) error {
+	namespace := scrt.GetNamespace()
+	saName := scrt.Annotations[ServiceAccountNameAnnotationKey]
+
+	chain, key, err := wc.GenKeyCertK8sCA(saName, namespace)
+	if err != nil {
+		return err
+	}
+
+	scrt.Data[CertChainID] = chain
+	scrt.Data[PrivateKeyID] = key
+	scrt.Data[RootCertID] = wc.getCurCACert()
+
+	_, err = wc.core.Secrets(namespace).Update(scrt)
+	return err
+}
+
+// Generate a certificate and key from k8s CA
+// Working flow:
+// 1. Generate a CSR
+// 2. Submit a CSR
+// 3. Approve a CSR
+// 4. Read the signed certificate
+// 5. Clean up the artifacts (e.g., delete CSR)
+func (wc *WebhookController) GenKeyCertK8sCA(saName string, saNamespace string) ([]byte, []byte, error) {
```

Comment:
Please move this to security/pkg/k8s.

Reply:
Done: moved to security/pkg/k8s/chiron/utils.go.